### PR TITLE
refactor(engine, tree): connect buffered blocks on pruner finish

### PR DESCRIPTION
**Pull request by shekhirin**
Two main changes:
1. Separate removing chains from blockchain tree and connecting buffered blocks to blockchain tree in `restore_canonical_hashes`, because we only need the buffered blocks part when pruner finishes.
1. Rename `restore_canonical_hashes` to `connect_buffered_blocks_to_canonical_hashes` and `restore_canonical_hashes_and_finalize` to `connect_buffered_blocks_to_canonical_hashes_and_finalize` accordingly.

### CODE REVIEW

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 1]
```diff
@@ -790,12 +782,44 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTree<DB, C, EF>
...
+        self.connect_buffered_blocks_to_hashes(last_canonical_hashes)?;
+
+        Ok(())
+    }
+
+    /// Reads the last `N` canonical hashes from the database and updates the block indices of the
+    /// tree by attempting to connect the buffered blocks to canonical hashes.
+    ///
+    /// `N` is the `max_reorg_depth` plus the number of block hashes needed to satisfy the
+    /// `BLOCKHASH` opcode in the EVM.
```
Comment (mattsse) [inline crates/blockchain-tree/src/blockchain_tree.rs:794]: I don't understand why we need this?

why do we need to store additional canonical hashes in the tree?

**Review by mattsse — COMMENTED** [review]
what does this fix exactly?

**Comment by shekhirin** [comment]
@mattsse when we finish the pruner, we need to connect the blocks buffered during the pruning. We've already been doing it, but also doing this https://github.com/paradigmxyz/reth/blob/458e3b98618b47c61e3766a1692e487a77f5307f/crates/blockchain-tree/src/blockchain_tree.rs#L775-L783 which is unnecessary, because canonical hashes in database didn't change while the pruner was running.

In case of scenario when the pipeline run is finished, we need to check the canonical hashes in database because they might've been changed: https://github.com/paradigmxyz/reth/blob/11f5f3f8d79eca8f03a6c714be338fc0c68f3faa/crates/consensus/beacon/src/engine/mod.rs#L1326-L1343

---

It wasn't a bug, we just did an unnecessary work and used one function for both cases (pipeline & pruner) which might be confusing.

File: crates/blockchain-tree/src/blockchain_tree.rs (hunk 1 above)
Comment (shekhirin) [reply crates/blockchain-tree/src/blockchain_tree.rs:794]: from my other PR, if I'm not mistaken: https://github.com/paradigmxyz/reth/blob/9fde694bbd026beeb16ee6739ca0190eeaa47635/crates/blockchain-tree/src/config.rs#L78-L82

**Review by mattsse — APPROVED** [review]
lgmt, I like this naming better

**Review by rakita — APPROVED** [review]
lgtm

**Pull request merged**